pex/auth.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import os
+from netrc import netrc
+
+from pex.compatibility import urlparse
+from pex.typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Iterable, Iterator, Optional, Text, Tuple
+
+    import attr  # vendor:skip
+else:
+    from pex.third_party import attr
+
+
+@attr.s(frozen=True)
+class PasswordEntry(object):
+    @staticmethod
+    def _reduced_url(url_info):
+        # type: (urlparse.ParseResult) -> str
+        return "{scheme}://{host}{port}".format(
+            scheme=url_info.scheme,
+            host=url_info.hostname,
+            port=":{port}".format(port=url_info.port) if url_info.port is not None else "",
+        )
+
+    @classmethod
+    def maybe_extract_from_url(cls, url):
+        # type: (str) -> Optional[PasswordEntry]
+        url_info = urlparse.urlparse(url)
+        if not url_info.username or not url_info.password:
+            return None
+
+        return cls(
+            uri=cls._reduced_url(url_info),
+            username=url_info.username,
+            password=url_info.password,
+        )
+
+    username = attr.ib()  # type: str
+    password = attr.ib()  # type: str
+    uri = attr.ib(default=None)  # type: Optional[str]
+
+    def uri_or_default(self, url):
+        # type: (Text) -> str
+        return self.uri or self._reduced_url(urlparse.urlparse(url))
+
+
+@attr.s(frozen=True)
+class PasswordDatabase(object):
+    @classmethod
+    def from_netrc(cls, netrc_file="~/.netrc"):
+        # type: (Optional[str]) -> PasswordDatabase
+        if not netrc_file:
+            return cls()
+
+        netrc_path = os.path.expanduser("~/.netrc")
+        if not os.path.isfile(netrc_path):
+            return cls()
+
+        def iter_entries():
+            # type: () -> Iterator[PasswordEntry]
+            for machine, (login, _, password) in netrc(netrc_path).hosts.items():
+                if password is None:
+                    continue
+
+                if machine == "default":
+                    # The 'default' entry is special and means just that; so we omit any qualifying
+                    # URI.
+                    yield PasswordEntry(username=login, password=password)
+
+                if urlparse.urlparse(machine).scheme:
```

Comment:
From my readings (e.g., https://www.labkey.org/Documentation/wiki-page.view?name=netrc, not sure how authoritative it is) it seems that in netrc a machine name is literally just a hostname, with no scheme or port. 

It also looks like the golang folks went through this exercise a couple of years ago (https://github.com/golang/go/issues/37130).  

So parsing a machine name as a URL seems like overkill? Although technically no harm. Just wondering if you'd encountered otherwise.  Perhaps worth documenting the thinking here (also see below for more perceived complications re ports).

Reply:
I did not encounter otherwise, I was just playing it safe. I'll document.